### ISSUE TITLE
TechEmpower: Failures with the queries and updates route

### ISSUE THREAD

**Issue by pushkarnk** [IBM-Swift/Kitura-Benchmarks#16]
In TechEmpower, the `queries` and `updates` routes fail with the `queries` parameters set to a value greater than 1 (say, `queries=5` or `queries=10`).

These are the failure messages with Kitura-Net:
```
[2018-11-28T09:46:41.649Z] [WARNING] [RouterResponse.swift:440 send] RouterResponse send(str:) invoked after end() for http://localhost/updates?queries=5
[2018-11-28T09:46:41.649Z] [WARNING] [RouterResponse.swift:185 end()] RouterResponse end() invoked more than once for http://localhost/updates?queries=5
[2018-11-28T09:46:43.058Z] [ERROR] [main.swift:164 TechEmpowerPostgres] OtherError("Timed out waiting for a DB connection from the pool")
[2018-11-28T09:46:43.058Z] [WARNING] [RouterResponse.swift:440 send] RouterResponse send(str:) invoked after end() for http://localhost/updates?queries=5
```

... and these are the failure messages with Kitura-NIO:
```
2018-11-28T10:08:40.909Z] [WARNING] [RouterResponse.swift:185 end()] RouterResponse end() invoked more than once for http://localhost:8080/updates?queries=5
[2018-11-28T10:08:40.909Z] [WARNING] [RouterResponse.swift:185 end()] RouterResponse end() invoked more than once for http://localhost:8080/updates?queries=5
[2018-11-28T10:08:40.908Z] [ERROR] [HTTPServerResponse.swift:123 end()] No channel available to end the response
[2018-11-28T10:08:40.907Z] [ERROR] [HTTPServerResponse.swift:90 write(from:)] No channel available to write
[2018-11-28T10:08:40.908Z] [ERROR] [HTTPServerResponse.swift:90 write(from:)] No channel available to write
[2018-11-28T10:08:40.907Z] [ERROR] [HTTPServerResponse.swift:90 write(from:)] No channel available to write
```

cc @djones6

**Comment by djones6** [comment]
The current implementation of the Kuery API is kind of a half-way house: we use async callbacks for DB execution but getting a connection from the connection pool is synchronous (and can block with a timeout).  Also, even though the Kuery API uses callbacks, it calls them synchronously, which means you can get away with things that shouldn't work, such as getting a connection and then performing two sequential `execute()` calls which are not nested.

@kilnerm and I have been redesigning this behaviour and the connection pool part of the API so that we do not block threads (although the underlying `libpq` calls still block).

I would try running with the latest asynchronous version of TechEmpower in this branch: https://github.com/IBM-Swift/Kitura-Benchmarks/tree/issue_async

This uses the `next` branches of Kuery and Kuery-Postgres which are 'more asynchronous'.

The call to `ConnectionPool.getConnection()` no longer blocks - it either returns a connection or enqueues the work to be done.  All callbacks are invoked on a different thread (giving the behaviour of asynchronous I/O, even though the underlying I/O is currently still blocking).  In the future, Kuery will use non-blocking I/O, but this will be a performance enhancement (not blocking threads internally) and should not result in a behavioural change for applications.

These changes will ultimately form part of [Swift-Kuery 3.0](https://github.com/IBM-Swift/Swift-Kuery/issues/162), and there are other changes in a similar vein in the pipeline:
- remaining synchronous parts of the Kuery API that are being converted to asynchronous callbacks: result fetching (ResultSet.fetchNext / asRows / titles)
- Swift-Kuery-ORM support for these changes.

Because the ORM has not yet been converted to use the changes in `next`, the ORM and Codable variants of the benchmarks are commented out in the issue_async branch.  I'll re-enable them once the ORM changes are ready.

**Comment by pushkarnk** [comment]
Error messages of this type are no longer seen:
```
[2018-11-28T10:08:40.909Z] [WARNING] [RouterResponse.swift:185 end()] RouterResponse end() invoked more than once for http://localhost:8080/updates?queries=5
```

However, we do have instances of this kind of error messages:
```
2018-11-28T10:08:40.907Z] [ERROR] [HTTPServerResponse.swift:90 write(from:)] No channel available to write
[2018-11-28T10:08:40.908Z] [ERROR] [HTTPServerResponse.swift:90 write(from:)] No channel available to write
[2018-11-28T10:08:40.907Z] [ERROR] [HTTPServerResponse.swift:90 write(from:)] No channel available to write
```

**Comment by pushkarnk** [comment]
The immediate source of these messages seems to be the fact the every `HTTPServerResponse` object keeps a weak reference to the `Channel`. When the `wrk` client exits, after its configured timeout duration, and closes its side of the connection, the `Channel` objects seem to get deallocated, causing a `nil Channel` reference in `HTTPServerResponse`. 

If `HTTPServerResponse` holds a strong reference to `Channel` these error messages go away. Logically, `HTTPServerResponse` and `Channel` must not for a retain cycle, and so may be fine to not have `Channel` as a weak var (can't recall why we made it weak in the first place). I quickly ran some valgrind tests and there's not memory leak.

**Comment by pushkarnk** [comment]
Created a PR: https://github.com/IBM-Swift/Kitura-NIO/pull/175

**Comment by pushkarnk** [comment]
This behaviour seems to be unique to the `wrk` client. I did a brief study of wrk and found that the socket close calls are issued asynchronously, meaning wrk does not wait for any response to be received before closing the sockets. Based on the `-d` value, the duration, the main thread in `wrk` sleeps for `duration` seconds, after which it wakes up and initiates the stoppage of every event loop. When the record_rate callback, which is initiated once per 100ms, executes the actual stopping of event loop happens, the sockets are closed in this process. 

Now on the server side, these socket closure events show up as read failures and BaseSocketChannel choses to close itself on read failures, this in turn shows up as the weak reference to Channel changing to nil and consequently [these error messages](https://github.com/IBM-Swift/Kitura-Benchmarks/issues/16#issuecomment-471410533) are reported. 

I think the error messages are legit and should be ignored. Kitura-NIO seems to be behaving in the right way.

**Comment by pushkarnk** [comment]
After a few discussions yesterday, I realise that the error messages are unwarranted and that https://github.com/IBM-Swift/Kitura-NIO/pull/175 is one of the ways of fixing them. Reopening the closed PR.

**Comment by pushkarnk** [comment]
Fixed by  IBM-Swift/Kitura-NIO#186 which is a lesser change than  IBM-Swift/Kitura-NIO#175
